src/algo.cpp

```diff
@@ -560,3 +560,445 @@ void auto_exposure_algorithm::histogram_score(std::vector<int>& h, const int tot
+
+    _corners[1].x = static_cast<double>(_pts[1].x - hs);
+    minimize_x(_ncc.data() + pos, _patch_size, f, _corners[1].x);
+
+    _corners[1].y = static_cast<double>(_pts[1].y - hs);
+    minimize_y(_ncc.data() + pos, _patch_size, f, _corners[1].y);
+
+    // lower left
+    pos = (_pts[2].y - hs) * _width + _pts[2].x - hs;
+
+    _corners[2].x = static_cast<double>(_pts[2].x - hs);
+    minimize_x(_ncc.data() + pos, _patch_size, f, _corners[2].x);
+
+    _corners[2].y = static_cast<double>(_pts[2].y - hs);
+    minimize_y(_ncc.data() + pos, _patch_size, f, _corners[2].y);
+
+    // lower right
+    pos = (_pts[3].y - hs) * _width + _pts[3].x - hs;
+
+    _corners[3].x = static_cast<double>(_pts[3].x - hs);
+    minimize_x(_ncc.data() + pos, _patch_size, f, _corners[3].x);
+
+    _corners[3].y = static_cast<double>(_pts[3].y - hs);
+    minimize_y(_ncc.data() + pos, _patch_size, f, _corners[3].y);
+}
+
+bool target_calculator::validate_corners(const uint8_t* img)
+{
+    uint8_t peaks[4] = { 0 };
+    int idx = 0;
+    int x = 0;
+    int y = 0;
+    for (int i = 0; i < 4; ++i)
+    {
+        y = static_cast<int>(_corners[i].y + 0.5f);
+        x = static_cast<int>(_corners[i].x + 0.5f);
+        idx = y * _width + x;
+        peaks[i] = img[idx];
+    }
+
+    static const int peak_diff_thresh = 12;
+    bool ok = true;
+    for (int j = 0; j < 4; ++j)
+    {
+        for (int i = 0; i < 4; ++i)
+        {
+            if (abs(peaks[i] - peaks[j]) > peak_diff_thresh)
+            {
+                ok = false;
+                break;
+            }
+
+            if (!ok)
+                break;
+        }
+    }
+
+    return ok;
+}
+
+void target_calculator::calculate_rect_sides(float rect_sides[4])
+{
+    double lx = _corners[1].x - _corners[0].x;
+    double ly = _corners[1].y - _corners[0].y;
+    rect_sides[0] = static_cast<float>(sqrt(lx * lx + ly * ly)); // uppper
+
+    lx = _corners[3].x - _corners[2].x;
+    ly = _corners[3].y - _corners[2].y;
+    rect_sides[1] = static_cast<float>(sqrt(lx * lx + ly * ly)); // lower
+
+    lx = _corners[2].x - _corners[0].x;
+    ly = _corners[2].y - _corners[0].y;
+    rect_sides[2] = static_cast<float>(sqrt(lx * lx + ly * ly)); // left
+
+    lx = _corners[3].x - _corners[1].x;
+    ly = _corners[3].y - _corners[1].y;
+    rect_sides[3] = static_cast<float>(sqrt(lx * lx + ly * ly)); // right
+}
+
+void target_calculator::minimize_x(const double* p, int s, double* f, double& x)
+{
+    int ws = _width - s;
+
+    for (int i = 0; i < s; ++i)
+        f[i] = 0;
+
+    for (int j = 0; j < s; ++j)
+    {
+        for (int i = 0; i < s; ++i)
+            f[i] += *p++;
+        p += ws;
+    }
+
+    x += subpixel_agj(f, s);
+}
+
+void target_calculator::minimize_y(const double* p, int s, double* f, double& y)
+{
+    int ws = _width - s;
+
+    for (int i = 0; i < s; ++i)
+        f[i] = 0;
+
+    for (int j = 0; j < s; ++j)
+    {
+        for (int i = 0; i < s; ++i)
```

Comment:
The line 937 can be move right before that one to avoid extra loop

Reply:
Please let me know the detailed suggestions here. Thanks!